File: packages/types/src/pipeline.ts
import type { PipelineType } from './node'

/** Serialized output from a single node after execution. */
export interface NodeOutputPreview {
  type: 'DataFrame' | 'Model' | 'Metrics' | 'Text' | 'Number' | 'Unknown'
  /** Variable name in the generated script */
  name: string
  // DataFrame
  shape?: [number, number]
  columns?: string[]
  dtypes?: Record<string, string>
  rows?: Record<string, unknown>[]
  // Model
  className?: string
  // Metrics / text / number
  value?: unknown
}

export interface NodeResult {
  /** All output variables captured after this node ran */
  outputs: NodeOutputPreview[]
  /** ms the node took to run */
  durationMs?: number
}

export interface PipelineNode {
  /** Instance ID (uuid) — unique within the pipeline */
  id: string
  /** References NodeDefinition.id */
  definitionId: string
  position: { x: number; y: number }
  /** User-filled values matching the node's configSchema */
  config: Record<string, unknown>
  /** Runtime state — set by executor */
  status?: 'idle' | 'running' | 'success' | 'error'
  error?: string
  /** Output preview captured after successful execution */
  result?: NodeResult
}

export interface PipelineEdge {
  id: string
  source: string
  sourceHandle: string
  target: string
  targetHandle: string
}

export interface PipelineDAG {
  id: string
  name: string
  pipeline: PipelineType
  nodes: PipelineNode[]
  edges: PipelineEdge[]
  createdAt: string
  updatedAt: string
}

export interface PipelineRun {
  id: string
  pipelineId: string
  status: 'pending' | 'running' | 'success' | 'failed'
  startedAt: string
  finishedAt?: string
  logs: string[]
}

File: apps/backend/app/models/pipeline.py
from __future__ import annotations
from typing import Any, Literal
from pydantic import BaseModel, Field


class NodePosition(BaseModel):
    x: float
    y: float


class PipelineNode(BaseModel):
    id: str
    definitionId: str
    position: NodePosition
    config: dict[str, Any] = Field(default_factory=dict)
    status: Literal["idle", "running", "success", "error"] = "idle"
    error: str | None = None


class PipelineEdge(BaseModel):
    id: str
    source: str
    sourceHandle: str = ""
    target: str
    targetHandle: str = ""


class PipelineDAG(BaseModel):
    id: str
    name: str
    pipeline: Literal["ml", "llm"]
    nodes: list[PipelineNode]
    edges: list[PipelineEdge]
    createdAt: str = ""
    updatedAt: str = ""


class PipelineRunRequest(BaseModel):
    dag: PipelineDAG
    env_vars: dict[str, str] = Field(default_factory=dict, description="Extra env vars injected into the pipeline subprocess (e.g. API keys)")


class PipelineRunResponse(BaseModel):
    run_id: str
    status: Literal["pending", "running", "success", "failed"]


class KubeflowRunRequest(BaseModel):
    dag: PipelineDAG
    kubeflow_host: str
    kubeflow_token: str = ""
    kubeflow_namespace: str = "kubeflow"
    experiment_name: str = "Default"
